File: typescript/src/lib/validation.ts
/**
 * Input validation utilities for Prompd CLI
 * Security-first validation following best practices
 */

import * as path from 'path';

/**
 * Custom error for security violations
 */
export class SecurityError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'SecurityError';
  }
}

/**
 * Custom error for validation failures
 */
export class ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ValidationError';
  }
}

/**
 * Validates npm-style package names
 * Rules: lowercase, alphanumeric, hyphens, dots, underscores
 * Scoped packages: @scope/name
 */
export function validatePackageName(name: string): boolean {
  if (!name || typeof name !== 'string') {
    return false;
  }

  // npm package name rules (scoped or unscoped)
  const pattern = /^(@[a-z0-9-~][a-z0-9-._~]*\/)?[a-z0-9-~][a-z0-9-._~]*$/;

  if (!pattern.test(name)) {
    return false;
  }

  // Length constraints
  if (name.length > 214) {
    return false;
  }

  // Cannot start with . or _
  const baseName = name.includes('/') ? name.split('/')[1] : name;
  if (baseName.startsWith('.') || baseName.startsWith('_')) {
    return false;
  }

  return true;
}

/**
 * Validates semantic versioning (x.y.z with optional pre-release/build metadata)
 */
export function validateVersion(version: string): boolean {
  if (!version || typeof version !== 'string') {
    return false;
  }

  // Semantic versioning pattern
  const semverPattern = /^\d+\.\d+\.\d+(-[a-zA-Z0-9.-]+)?(\+[a-zA-Z0-9.-]+)?$/;
  return semverPattern.test(version);
}

/**
 * Validates namespace/scope format (@namespace)
 */
export function validateNamespace(namespace: string): boolean {
  if (!namespace || typeof namespace !== 'string') {
    return false;
  }

  // Must start with @
  if (!namespace.startsWith('@')) {
    return false;
  }

  const scope = namespace.slice(1);

  // npm scope rules: lowercase, URL-safe characters
  return /^[a-z0-9-~][a-z0-9-._~]*$/.test(scope);
}

/**
 * Validates registry URL
 * Only HTTPS allowed (except localhost for development)
 */
export function validateRegistryUrl(url: string): boolean {
  if (!url || typeof url !== 'string') {
    return false;
  }

  try {
    const parsed = new URL(url);

    // Only allow HTTPS (except localhost for dev)
    if (parsed.protocol !== 'https:') {
      // Allow HTTP only for localhost
      if (parsed.hostname !== 'localhost' && parsed.hostname !== '127.0.0.1') {
        return false;
      }
    }

    // Don't allow data:, file:, javascript:, etc.
    if (!['http:', 'https:'].includes(parsed.protocol)) {
      return false;
    }

    return true;
  } catch {
    return false;
  }
}

/**
 * Sanitizes file paths to prevent directory traversal
 * Throws SecurityError if path is dangerous
 */
export function sanitizeFilePath(filePath: string, basePath?: string): string {
  if (!filePath || typeof filePath !== 'string') {
    throw new ValidationError('Invalid file path');
  }

  // Normalize the path
  const normalized = path.normalize(filePath);

  // Check for directory traversal attempts
  if (normalized.includes('..')) {
    throw new SecurityError('Directory traversal detected in file path');
  }

  // Check for absolute paths (when we expect relative)
  if (path.isAbsolute(normalized) && basePath) {
    throw new SecurityError('Absolute paths not allowed');
  }

  // If basePath provided, ensure the path stays within it
  if (basePath) {
    const resolvedPath = path.resolve(basePath, normalized);
    const resolvedBase = path.resolve(basePath);

    if (!resolvedPath.startsWith(resolvedBase + path.sep) &&
        resolvedPath !== resolvedBase) {
      throw new SecurityError('Path escapes base directory');
    }

    return resolvedPath;
  }

  return normalized;
}

/**
 * Validates file extension against whitelist
 */
export function validateFileExtension(filePath: string, allowedExtensions: string[]): boolean {
  if (!filePath || typeof filePath !== 'string') {
    return false;
  }

  if (!Array.isArray(allowedExtensions) || allowedExtensions.length === 0) {
    return false;
  }

  const ext = path.extname(filePath).toLowerCase();
  return allowedExtensions.includes(ext);
}

/**
 * Validates command-line flag/option names
 */
export function validateOptionName(name: string): boolean {
  if (!name || typeof name !== 'string') {
    return false;
  }

  // Allow letters, numbers, hyphens (typical CLI option names)
  return /^[a-zA-Z][a-zA-Z0-9-]*$/.test(name);
}

/**
 * Validates parameter keys (for .prmd parameters)
 */
export function validateParameterKey(key: string): boolean {
  if (!key || typeof key !== 'string') {
    return false;
  }

  // Allow letters, numbers, underscores (snake_case or camelCase)
  if (!/^[a-zA-Z_][a-zA-Z0-9_]*$/.test(key)) {
    return false;
  }

  // Length limits
  if (key.length > 128) {
    return false;
  }

  return true;
}

/**
 * Sanitizes parameter keys by converting to safe format
 */
export function sanitizeParameterKey(key: string): string {
  if (typeof key !== 'string') {
    throw new ValidationError('Parameter key must be string');
  }

  // Remove potentially dangerous characters, replace with underscore
  const sanitized = key.replace(/[^a-zA-Z0-9_]/g, '_');

  if (sanitized.length === 0 || sanitized.length > 128) {
    throw new ValidationError('Invalid parameter key length');
  }

  return sanitized;
}

/**
 * Detects potential secrets in strings
 * Returns array of detected secret patterns
 */
export function detectSecrets(content: string): Array<{ type: string; match: string; line: number }> {
  const secrets: Array<{ type: string; match: string; line: number }> = [];

  // Split into lines for line number tracking
  const lines = content.split('\n');

  // Secret patterns to detect
  const patterns = [
    { type: 'OpenAI API Key', regex: /sk-[a-zA-Z0-9]{32,}/g },
    { type: 'Anthropic API Key', regex: /sk-ant-[a-zA-Z0-9-_]{32,}/g },
    { type: 'Prompd Registry Token', regex: /prompd_[a-zA-Z0-9-_]{32,}/g },
    { type: 'Generic API Key', regex: /api[_-]?key["\s:=]+[a-zA-Z0-9-_]{20,}/gi },
    { type: 'Generic Secret', regex: /secret["\s:=]+[a-zA-Z0-9-_]{20,}/gi },
    { type: 'Private Key Header', regex: /-----BEGIN (RSA |EC |DSA )?PRIVATE KEY-----/g },
    { type: 'AWS Access Key', regex: /AKIA[0-9A-Z]{16}/g },
    { type: 'GitHub Token', regex: /gh[pousr]_[a-zA-Z0-9]{36,}/g },
  ];

  lines.forEach((line, index) => {
    patterns.forEach(({ type, regex }) => {
      const matches = line.match(regex);
      if (matches) {
        matches.forEach((match) => {
          // Mask the secret in the output (show first 8 chars only)
          const maskedMatch = match.length > 8
            ? match.slice(0, 8) + '...'
            : '***';

          secrets.push({
            type,
            match: maskedMatch,
            line: index + 1
          });
        });
      }
    });
  });

  return secrets;
}

/**
 * Validates git reference (tag, branch, commit hash)
 */
export function validateGitRef(ref: string): boolean {
  if (!ref || typeof ref !== 'string') {
    return false;
  }

  // Allow alphanumeric, dots, hyphens, slashes, underscores
  // Covers: branch names, tags, commit hashes
  return /^[a-zA-Z0-9._/-]+$/.test(ref);
}

/**
 * Validates model name (for LLM providers)
 */
export function validateModelName(name: string): boolean {
  if (!name || typeof name !== 'string') {
    return false;
  }

  // Allow letters, numbers, hyphens, dots, underscores
  // e.g., gpt-4, claude-3-opus-20240229, llama3.2
  return /^[a-zA-Z0-9._-]+$/.test(name);
}

/**
 * Validates provider name
 */
export function validateProviderName(name: string): boolean {
  if (!name || typeof name !== 'string') {
    return false;
  }

  // Lowercase letters, numbers, hyphens only
  return /^[a-z][a-z0-9-]*$/.test(name);
}

/**
 * Validates email address (basic validation)
 */
export function validateEmail(email: string): boolean {
  if (!email || typeof email !== 'string') {
    return false;
  }

  // Basic email pattern
  const emailPattern = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
  return emailPattern.test(email);
}

/**
 * Checks if a string contains potentially malicious content
 */
export function containsMaliciousContent(content: string): boolean {
  if (!content || typeof content !== 'string') {
    return false;
  }

  // Patterns that might indicate injection attempts
  const maliciousPatterns = [
    /<script[^>]*>/i,           // Script tags
    /javascript:/i,              // JavaScript protocol
    /on\w+\s*=/i,               // Event handlers (onclick, onerror, etc.)
    /eval\s*\(/i,               // eval calls
    /Function\s*\(/i,           // Function constructor
    /\$\{.*\}/,                 // Template literal injection
    /`.*`/,                     // Backticks (potential template literals)
  ];

  return maliciousPatterns.some(pattern => pattern.test(content));
}

/**
 * Validates that a value is within numeric range
 */
export function validateNumericRange(value: number, min: number, max: number): boolean {
  if (typeof value !== 'number' || !isFinite(value)) {
    return false;
  }

  return value >= min && value <= max;
}

/**
 * Validates string length
 */
export function validateStringLength(value: string, minLength: number, maxLength: number): boolean {
  if (!value || typeof value !== 'string') {
    return false;
  }

  return value.length >= minLength && value.length <= maxLength;
}

/**
 * Sanitizes string for safe display (prevents injection in terminal output)
 */
export function sanitizeForDisplay(value: string): string {
  if (!value || typeof value !== 'string') {
    return '';
  }

  // Remove control characters except common whitespace
  return value.replace(/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F-\x9F]/g, '');
}

/**
 * Validates JSON structure
 */
export function validateJSON(jsonString: string): boolean {
  try {
    JSON.parse(jsonString);
    return true;
  } catch {
    return false;
  }
}

/**
 * Validates YAML structure
 */
export function validateYAML(yamlString: string): boolean {
  try {
    // This would require yaml library - just check basic structure
    // In real implementation, use js-yaml to parse
    return typeof yamlString === 'string' && yamlString.length > 0;
  } catch {
    return false;
  }
}

File: go/cmd/prompd/validation.go
package main

import (
	"errors"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// validatePackageName validates package name format (npm-style scoped packages)
func validatePackageName(name string) error {
	if name == "" {
		return errors.New("package name cannot be empty")
	}

	// Must start with @ for scoped packages
	if !strings.HasPrefix(name, "@") {
		return errors.New("package name must be scoped (start with @)")
	}

	// Format: @namespace/package-name
	// Allow lowercase letters, numbers, hyphens
	pattern := regexp.MustCompile(`^@[a-z0-9][a-z0-9-]*\/[a-z0-9][a-z0-9-]*$`)
	if !pattern.MatchString(name) {
		return errors.New("invalid package name format (expected @namespace/package-name with lowercase letters, numbers, and hyphens)")
	}

	// Check for dangerous patterns
	if strings.Contains(name, "..") || strings.Contains(name, "//") {
		return errors.New("package name contains invalid characters")
	}

	// Check length constraints
	if len(name) > 214 {
		return errors.New("package name too long (max 214 characters)")
	}

	parts := strings.Split(name, "/")
	if len(parts) != 2 {
		return errors.New("package name must have exactly one '/' separator")
	}

	namespace := parts[0] // includes @
	pkgName := parts[1]

	if len(namespace) < 2 {
		return errors.New("namespace too short")
	}

	if len(pkgName) < 1 {
		return errors.New("package name too short")
	}

	return nil
}

// validateVersion validates semantic versioning format
func validateVersion(version string) error {
	if version == "" {
		return errors.New("version cannot be empty")
	}

	// Strict semver: x.y.z (numbers only)
	pattern := regexp.MustCompile(`^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$`)
	if !pattern.MatchString(version) {
		return errors.New("version must follow semantic versioning format (x.y.z with non-negative integers)")
	}

	// Additional check: parse and validate ranges
	parts := strings.Split(version, ".")
	if len(parts) != 3 {
		return errors.New("version must have exactly 3 parts (major.minor.patch)")
	}

	return nil
}

// validateRegistryURL validates registry URL format and security
func validateRegistryURL(urlStr string) error {
	if urlStr == "" {
		return errors.New("URL cannot be empty")
	}

	parsedURL, err := url.Parse(urlStr)
	if err != nil {
		return fmt.Errorf("invalid URL format: %v", err)
	}

	// Must be HTTPS except for localhost
	if parsedURL.Scheme != "https" {
		// Allow HTTP only for localhost and 127.0.0.1
		if !strings.Contains(parsedURL.Host, "localhost") &&
			!strings.Contains(parsedURL.Host, "127.0.0.1") &&
			!strings.Contains(parsedURL.Host, "[::1]") {
			return errors.New("registry URL must use HTTPS (except localhost for development)")
		}
	}

	// Must have host
	if parsedURL.Host == "" {
		return errors.New("registry URL must have a host")
	}

	// Check for dangerous patterns in URL
	if strings.Contains(urlStr, "..") || strings.Contains(urlStr, "//") {
		// Allow // in scheme (https://) but not elsewhere
		withoutScheme := strings.TrimPrefix(urlStr, parsedURL.Scheme+"://")
		if strings.Contains(withoutScheme, "//") || strings.Contains(withoutScheme, "..") {
			return errors.New("registry URL contains suspicious patterns")
		}
	}

	return nil
}

// sanitizeFilePath sanitizes file paths to prevent directory traversal
func sanitizeFilePath(path string, basePath string) (string, error) {
	// SECURITY: Check for null bytes on raw input BEFORE cleaning
	if strings.Contains(path, "\x00") {
		return "", errors.New("path contains null bytes")
	}

	// Clean the path
	cleanPath := filepath.Clean(path)

	// Check for absolute paths (should be relative)
	if filepath.IsAbs(cleanPath) {
		return "", errors.New("path must be relative, not absolute")
	}

	// Check for dangerous patterns
	if strings.HasPrefix(cleanPath, "..") ||
		strings.Contains(cleanPath, string(filepath.Separator)+"..") ||
		strings.Contains(cleanPath, ".."+string(filepath.Separator)) {
		return "", errors.New("path contains directory traversal (..)  ")
	}

	// If basePath provided, ensure path is within it
	if basePath != "" {
		absPath, err := filepath.Abs(filepath.Join(basePath, cleanPath))
		if err != nil {
			return "", fmt.Errorf("could not resolve absolute path: %v", err)
		}

		absBase, err := filepath.Abs(basePath)
		if err != nil {
			return "", fmt.Errorf("could not resolve base path: %v", err)
		}

		// Ensure absPath starts with absBase
		if !strings.HasPrefix(absPath, absBase) {
			return "", errors.New("path escapes base directory")
		}

		return cleanPath, nil
	}

	return cleanPath, nil
}

// isSecurePath validates that a path is safe for ZIP operations (enhanced ZIP slip protection)
func isSecurePath(path string, basePath string) error {
	// SECURITY: Check for null bytes on raw input BEFORE cleaning
	if strings.Contains(path, "\x00") {
		return errors.New("path contains null bytes")
	}

	// Clean and normalize path
	cleanPath := filepath.Clean(path)

	// Check for absolute paths
	if filepath.IsAbs(cleanPath) {
		return errors.New("path is absolute (must be relative)")
	}

	// Check for directory traversal
	if strings.HasPrefix(cleanPath, "..") ||
		strings.Contains(cleanPath, string(filepath.Separator)+"..") ||
		strings.Contains(cleanPath, ".."+string(filepath.Separator)) {
		return errors.New("path contains directory traversal")
	}

	// Check for symlinks (if file exists)
	if basePath != "" {
		fullPath := filepath.Join(basePath, cleanPath)
		if info, err := os.Lstat(fullPath); err == nil {
			if info.Mode()&os.ModeSymlink != 0 {
				return errors.New("path is a symlink (security risk)")
			}
		}
	}

	// Validate final path is within base
	if basePath != "" {
		finalPath, err := filepath.Abs(filepath.Join(basePath, cleanPath))
		if err != nil {
			return fmt.Errorf("could not resolve final path: %v", err)
		}

		absBase, err := filepath.Abs(basePath)
		if err != nil {
			return fmt.Errorf("could not resolve base path: %v", err)
		}

		// Ensure finalPath is within absBase (directory traversal check)
		if !strings.HasPrefix(finalPath+string(filepath.Separator), absBase+string(filepath.Separator)) {
			return errors.New("path escapes base directory")
		}
	}

	return nil
}

// isMaliciousContent checks for potentially malicious code patterns
func isMaliciousContent(content string) bool {
	maliciousPatterns := []string{
		"eval(",
		"exec(",
		"os.system(",
		"subprocess.call(",
		"subprocess.Popen(",
		"Runtime.getRuntime().exec(",
		"<script>",
		"javascript:",
		"cmd /c",
		"powershell -",
		"base64 -d",
		"chmod +x",
		"/bin/sh",
		"/bin/bash",
	}

	lowerContent := strings.ToLower(content)
	for _, pattern := range maliciousPatterns {
		if strings.Contains(lowerContent, strings.ToLower(pattern)) {
			return true
		}
	}

	return false
}

// validateDescription validates package description
func validateDescription(description string) error {
	if description == "" {
		return errors.New("description cannot be empty")
	}

	if len(description) > 500 {
		return errors.New("description too long (max 500 characters)")
	}

	// Check for suspicious patterns
	if isMaliciousContent(description) {
		return errors.New("description contains suspicious content")
	}

	return nil
}

// validateAuthor validates package author field
func validateAuthor(author string) error {
	if author == "" {
		return nil // Author is optional
	}

	if len(author) > 200 {
		return errors.New("author field too long (max 200 characters)")
	}

	// Check for suspicious patterns
	if isMaliciousContent(author) {
		return errors.New("author field contains suspicious content")
	}

	return nil
}
